channeldb/invoices.go

```diff
@@ -1332,7 +1332,112 @@ func (d *DB) updateInvoice(hash lntypes.Hash, invoices, settleIndex *bbolt.Bucke
 	return &invoice, nil
 }
 
-func setSettleFields(settleIndex *bbolt.Bucket, invoiceNum []byte,
+// updateInvoiceState validates and processes an invoice state update.
+func updateInvoiceState(invoice *Invoice, hash lntypes.Hash,
+	update InvoiceStateUpdateDesc) error {
+
+	switch update.NewState {
+
+	case ContractOpen:
+		return errors.New("cannot move invoice to open")
+
+	case ContractAccepted:
+		if invoice.State != ContractOpen {
+			return errors.New(
+				"can only accept open invoices")
+		}
+
+	case ContractSettled:
+		// Validate preimage.
+		if update.Preimage.Hash() != hash {
+			return errors.New("preimage does not match")
```

Comment:
Define these errors as done for `ErrInvoiceAlreadyAccepted` and update tests to check specific error while you're here?

Reply:
Errors defined. But where did you see these errors being checked on?

Reply:
This is also a more fundamental issue. When to define an error var and when just return it in-line? In other places in this file, we still create in-line errors. Some of which also have parameters. That can all be structured too, but without that specific error being matched, it may be just boiler plate?